text/0036-async-testbench-functions.md

```diff
@@ -0,0 +1,235 @@
+        await sim.negedge()
+        await sim.set(pin, o[0])
+        await sim.posedge()
+        await sim.set(pin, o[1])
+sim.add_process(ddr_buffer)
+```
+
+Flop with configurable edge reset and posedge clock as a process:
+```python
+clk = Signal(); rst = Signal(); d = Signal(); q = Signal()
+def dff(rst_edge):
+    async def process(sim):
+        async for clk_val, rst_val in sim.posedge(clk).edge(rst, rst_edge):
+            await sim.set(q, 0 if rst_val == rst_edge else await sim.get(d))
+    return process
+sim.add_process(dff(rst_edge=0))
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following `Simulator` methods have their signatures updated:
+
+* `add_process(process)`
+* `add_testbench(process, *, background=False)`
+
+The new optional named argument `background` registers the testbench as a background process when true.
+
+Both methods are updated to accept an async function passed as `process`.
+The async function must accept an argument `sim`, which will be passed a simulator context.
+(Argument name is just convention, will be passed positionally.)
+
+The simulator context have the following methods:
+- `get(expr: Value) -> int`
+- `get(expr: ValueCastable) -> any`
+  - Returns the value of `expr` when awaited.
+    When `expr` is a value-castable, the value will be converted through `.from_bits()`.
+- `set(expr: Value, value: ConstLike)`
+- `set(expr: ValueCastable, value: any)`
+  - Set `expr` to `value` when awaited.
+    When `expr` is a value-castable, the value will be converted through `.const()`.
+- `memory_read(instance: MemoryInstance, address)`
+  - Read the value from `address` in `instance` when awaited.
+- `memory_write(instance: MemoryInstance, address, value, mask=None)`
+  - Write `value` to `address` in `instance` when awaited. If `mask` is given, only the corresponding bits are written.
+- `delay(interval)`
+  - Create a trigger object for advancing simulation by `interval` seconds.
+- `tick(domain="sync", *, context=None)`
+  - Create a trigger object for advancing simulation by one tick of `domain`.
+    When an elaboratable is passed to `context`, `domain` will be resolved from its perspective.
+  - If `domain` is asynchronously reset while this is being awaited, `AsyncReset` is raised.
+- `changed(*signals)`
+  - Create a trigger object for advancing simulation until any signal in `signals` changes.
+- `edge(signal, value)`
+  - Create a trigger object for advancing simulation until `signal` is changed to `value`.
+    `signal` must be a 1-bit signal or a 1-bit slice of a signal.
+- `posedge(signal)`
+- `negedge(signal)`
+  - Aliases for `edge(signal, 1)` and `edge(signal, 0)` respectively.
+  
+  `signal` is changed to `value`. `None` is a wildcard and will trigger on any change.
+- `critical()`
+  - Return a context manager that ensures simulation won't terminate in the middle of the enclosed scope.
```

Comment:
That ensures simulation won't terminate *because run() ran out of non-background processes*. I think this may be better put as "if the current process is a background process, `async with critical()` makes it a non-background process for the duration of the statement".